### PR TITLE
[khan academy] Fix extractor

### PR DESCRIPTION
**Pull request by rand-net**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Updates khan academy api.

### CODE REVIEW

**Comment by rand-net** [comment]
In the last PR, I thought the online github editor was flake compliant and I hard coded the api url.

**Comment by pukkandan** [comment]
> In the last PR, I thought the online github editor was flake compliant and I hard coded the api url.

FYI, PRs are linked to the branch from which it is created. You can push any change to the branch, and it will be reflected in the PR

File: yt_dlp/extractor/khanacademy.py [hunk 1]
```diff
@@ -25,15 +25,39 @@ def _parse_video(self, video):
...
+            + country_code
+            + '%22%7D'
+        )
+
+        component_props = self._parse_json(
+            self._download_json(query_url, display_id,)[
+                'data'
+            ]['contentJson'],
+            display_id,
+        )['componentProps']
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:60]: Use query parameter of _download_json

File: yt_dlp/extractor/khanacademy.py [hunk 2]
```diff
@@ -25,15 +25,39 @@ def _parse_video(self, video):
...
-                    'queryParams': '',
-                }),
-            })['data']['contentJson'], display_id)['componentProps']
+        fastly_cacheable = 'persist_until_publish'
+        query_hash = 4134764944
+        query_lang = 'en'
+        query_params = 'lang=en'
+        is_modal = 'false'
+        follow_redirects = 'true'
+        country_code = 'US'
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:34]: unnecessary variables

File: yt_dlp/extractor/khanacademy.py [hunk 3]
```diff
@@ -9,7 +7,9 @@
 
 
 class KhanAcademyBaseIE(InfoExtractor):
-    _VALID_URL_TEMPL = r'https?://(?:www\.)?khanacademy\.org/(?P<id>(?:[^/]+/){%s}%s[^?#/&]+)'
+    _VALID_URL_TEMPL = (
+        r'https?://(?:www\.)?khanacademy\.org/(?P<id>(?:[^/]+/){%s}%s[^?#/&]+)'
+    )
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:12]: why?
```suggestion
    _VALID_URL_TEMPL = r'https?://(?:www\.)?khanacademy\.org/(?P<id>(?:[^/]+/){%s}%s[^?#/&]+)'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
> - [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)

Clearly not

**Comment by pukkandan** [comment]
What does this fix anyway?

**Comment by rand-net** [comment]
> What does this fix anyway?

I updated the broken API. Here is the error log, testing with the latest version.

`
$ yt-dlp "https://www.khanacademy.org/math/algebra/x2f8bb11595b61c86:foundation-algebra"         

[khanacademy:unit] math/algebra/x2f8bb11595b61c86:foundation-algebra: Downloading JSON metadata
ERROR: [khanacademy:unit] math/algebra/x2f8bb11595b61c86:foundation-algebra: Unable to download JSON metadata: HTTP Error 400: Bad Request (caused by <HTTPError 400: 'Bad Request'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
`

File: yt_dlp/extractor/khanacademy.py [hunk 4]
```diff
@@ -25,15 +23,23 @@ def _parse_video(self, video):
...
+                display_id,
+                query={
+                    'fastly_cacheable': ['persist_until_publish'],
+                    'hash': ['4134764944'],
+                    'lang': ['en'],
+                    'variables': [
+                        '{"path":"'
+                        + display_id
+                        + '","queryParams":"lang=en","isModal":false,"followRedirects":true,"countryCode":"US"}'
+                    ],
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:38]:
```py
                    'variables': json.dumps({
                        'path': display_id,
                        ...,
                    }),
```

File: yt_dlp/extractor/khanacademy.py [hunk 5]
```diff
@@ -62,11 +68,13 @@ def _parse_component_props(self, component_props):
...
-            'timestamp': parse_iso8601(video.get('dateAdded')),
-            'license': video.get('kaUserLicense'),
-        })
+        info.update(
+            {
+                'uploader': ', '.join(author_names) if author_names else None,
+                'timestamp': parse_iso8601(video.get('dateAdded')),
+                'license': video.get('kaUserLicense'),
+            }
+        )
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:77]: unnecessary change

File: yt_dlp/extractor/khanacademy.py [hunk 6]
```diff
@@ -87,19 +95,25 @@ def _parse_component_props(self, component_props):
         curation = component_props['curation']
 
         entries = []
-        tutorials = try_get(curation, lambda x: x['tabs'][0]['modules'][0]['tutorials'], list) or []
+        tutorials = (
+            try_get(curation, lambda x: x['tabs'][0]['modules'][0]['tutorials'], list)
+            or []
+        )
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:101]: same

File: yt_dlp/extractor/khanacademy.py [hunk 7]
```diff
@@ -87,19 +95,25 @@ def _parse_component_props(self, component_props):
...
                     entries.append(info)
 
         return self.playlist_result(
-            entries, curation.get('unit'), curation.get('title'),
-            curation.get('description'))
+            entries,
+            curation.get('unit'),
+            curation.get('title'),
+            curation.get('description'),
+        )
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:119]: same

File: yt_dlp/extractor/khanacademy.py [hunk 8]
```diff
@@ -25,15 +23,23 @@ def _parse_video(self, video):
...
+                    'variables': [
+                        '{"path":"'
+                        + display_id
+                        + '","queryParams":"lang=en","isModal":false,"followRedirects":true,"countryCode":"US"}'
+                    ],
+                },
+            )['data']['contentJson'],
+            display_id,
+        )['componentProps']
         return self._parse_component_props(component_props)
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:43]:
```py
        content = self._download_json(
            ...)['data']['contentJson']
        return self._parse_component_props(self._parse_json(content, display_id)['componentProps'])
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
> > - [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)

Clearly not. The whole thing needs to be refactored accordingly

I see you still can't be bothered with this!

File: yt_dlp/extractor/khanacademy.py (hunk 5 above)
Comment (rand-net) [reply yt_dlp/extractor/khanacademy.py:77]: Got it. I will turn off black auto-formatting.

File: yt_dlp/extractor/khanacademy.py [hunk 9]
```diff
@@ -25,16 +25,21 @@ def _parse_video(self, video):
...
         display_id = self._match_id(url)
-        component_props = self._parse_json(self._download_json(
-            'https://www.khanacademy.org/api/internal/graphql',
+        content = self._download_json(
+            'https://www.khanacademy.org/api/internal/graphql/FetchContentData',
             display_id, query={
-                'hash': 1604303425,
+                'fastly_cacheable': ['persist_until_publish'],
+                'hash': ['4134764944'],
+                'lang': ['en'],
```
Comment (pukkandan) [inline yt_dlp/extractor/khanacademy.py:33]:
```suggestion
                'fastly_cacheable': 'persist_until_publish',
                'hash': '4134764944',
                'lang': 'en',
```
